src/lib.rs

```diff
@@ -1,41 +1,278 @@
 #![no_std]
 
-use gstd::{debug, msg, prelude::*};
+pub mod messages;
+pub use messages::*;
 
-#[no_mangle]
-pub unsafe extern "C" fn handle() {
-    debug!("handle()");
-    let payload = String::from_utf8(msg::load_bytes()).expect("Invalid handle message");
+pub mod asserts;
+pub use asserts::*;
+
+use ico_io::*;
+
+use gstd::{prelude::*, exec, msg, ActorId, debug}; 
+
+#[derive(Default)]
+struct IcoContract {
+    ico_state: IcoState,
+    start_price: u128,
+    price_increase_step: u128,
+    time_increase_step: u128,
+    tokens_sold: u128,
+    tokens_goal: u128,
+    owner: ActorId, 
+    token_id: ActorId,
+    token_holders: BTreeMap<ActorId, u128>,
+}
+
+static mut ICO_CONTRACT: Option<IcoContract> = None;
+
+impl IcoContract {
 
-    if payload == "Hello" {
-        msg::reply(b"World", 0).unwrap();
+    /// Starts ICO contract
+    /// 
+    /// Requirements:
+    /// * Only owner can start ICO
+    /// * At least `tokens_goal` tokens need to be minted
+    /// * ICO can be started only once
+    /// * All arguments must be greater than zero
+    /// 
+    /// Arguments:
+    /// * `config`: Consists of `duration`, `start_price`, `tokens_goal`, `price_increase_step` and time_increase_step 
+    /// 
+    async fn start_ico(&mut self, config: IcoAction) {
+        check_input(&config);
+        assert_owner_message(&self.owner, "start_ico(): Not owner start ICO");
```

Comment:
It is not an idiomatic way to call extern functions: https://doc.rust-lang.org/stable/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html
```suggestion
        asserts::owner_message(&self.owner, "start_ico(): Not owner starts ICO");
```